File: tools/load_test.py
import http.client
import time
import json
import sys
import statistics
import concurrent.futures
import urllib.parse

BASE_HOST = "localhost"
BASE_PORT = 5000
SEARCH_PATH = "/user/search?first_name=Ал&last_name=Ив"
DURATION_SECONDS = 10

def make_request():
    start = time.perf_counter()
    try:
        conn = http.client.HTTPConnection(BASE_HOST, BASE_PORT, timeout=30)
        conn.request("GET", SEARCH_PATH)
        resp = conn.getresponse()
        resp.read()
        status = resp.status
        conn.close()
        elapsed = time.perf_counter() - start
        return elapsed, status
    except Exception as e:
        elapsed = time.perf_counter() - start
        return elapsed, 0

def run_load_test(concurrency, duration=DURATION_SECONDS):
    results = []
    errors = 0
    stop_time = time.perf_counter() + duration
    total_requests = 0

    with concurrent.futures.ThreadPoolExecutor(max_workers=concurrency) as executor:
        futures = []
        while time.perf_counter() < stop_time:
            if len(futures) < concurrency * 2:
                futures.append(executor.submit(make_request))
            
            done = [f for f in futures if f.done()]
            for f in done:
                elapsed, status = f.result()
                results.append(elapsed)
                if status != 200:
                    errors += 1
                total_requests += 1
                futures.remove(f)

        for f in concurrent.futures.as_completed(futures, timeout=30):
            elapsed, status = f.result()
            results.append(elapsed)
            if status != 200:
                errors += 1
            total_requests += 1

    if not results:
        return {}

    results_ms = [r * 1000 for r in results]
    results_ms.sort()

    return {
        "concurrency": concurrency,
        "total_requests": len(results_ms),
        "duration_sec": duration,
        "throughput_rps": len(results_ms) / duration,
        "avg_ms": statistics.mean(results_ms),
        "median_ms": statistics.median(results_ms),
        "p95_ms": results_ms[int(len(results_ms) * 0.95)] if len(results_ms) > 20 else max(results_ms),
        "p99_ms": results_ms[int(len(results_ms) * 0.99)] if len(results_ms) > 100 else max(results_ms),
        "max_ms": max(results_ms),
        "min_ms": min(results_ms),
        "errors": errors
    }

if __name__ == "__main__":
    label = sys.argv[1] if len(sys.argv) > 1 else "test"
    concurrency_levels = [1, 10, 100, 1000]
    all_results = []

    for c in concurrency_levels:
        print(f"Running with concurrency={c} for {DURATION_SECONDS}s...")
        result = run_load_test(c)
        all_results.append(result)
        print(f"  Requests: {result['total_requests']}, "
              f"RPS: {result['throughput_rps']:.1f}, "
              f"Avg: {result['avg_ms']:.1f}ms, "
              f"P95: {result['p95_ms']:.1f}ms, "
              f"P99: {result['p99_ms']:.1f}ms, "
              f"Errors: {result['errors']}")

    output_file = f"load_test_{label}.json"
    with open(output_file, "w") as f:
        json.dump(all_results, f, indent=2)
    print(f"\nResults saved to {output_file}")

File: tools/load_test.js
const http = require("http");
const fs = require("fs");

const BASE_URL = "http://localhost:5000/user/search?first_name=%D0%90%D0%BB&last_name=%D0%98%D0%B2";
const DURATION_MS = 10_000;

function makeRequest() {
  return new Promise((resolve) => {
    const start = performance.now();
    const req = http.get(BASE_URL, { timeout: 30000 }, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => resolve({ elapsed: performance.now() - start, status: res.statusCode }));
    });
    req.on("error", () => resolve({ elapsed: performance.now() - start, status: 0 }));
    req.on("timeout", () => { req.destroy(); resolve({ elapsed: performance.now() - start, status: 0 }); });
  });
}

async function runWorker(results, stopTime) {
  while (performance.now() < stopTime) {
    const r = await makeRequest();
    results.push(r);
  }
}

async function runLoadTest(concurrency) {
  const results = [];
  const stopTime = performance.now() + DURATION_MS;
  const workers = [];
  for (let i = 0; i < concurrency; i++) workers.push(runWorker(results, stopTime));
  await Promise.all(workers);

  const latencies = results.map((r) => r.elapsed).sort((a, b) => a - b);
  const errors = results.filter((r) => r.status !== 200).length;
  const sum = latencies.reduce((a, b) => a + b, 0);

  return {
    concurrency,
    total_requests: latencies.length,
    duration_sec: DURATION_MS / 1000,
    throughput_rps: +(latencies.length / (DURATION_MS / 1000)).toFixed(1),
    avg_ms: +(sum / latencies.length).toFixed(1),
    median_ms: +latencies[Math.floor(latencies.length * 0.5)].toFixed(1),
    p95_ms: +latencies[Math.floor(latencies.length * 0.95)].toFixed(1),
    p99_ms: +latencies[Math.floor(latencies.length * 0.99)].toFixed(1),
    max_ms: +latencies[latencies.length - 1].toFixed(1),
    min_ms: +latencies[0].toFixed(1),
    errors,
  };
}

async function main() {
  const label = process.argv[2] || "test";
  const levels = [1, 10, 100, 1000];
  const allResults = [];

  for (const c of levels) {
    process.stdout.write(`Running concurrency=${c} for ${DURATION_MS / 1000}s... `);
    const r = await runLoadTest(c);
    allResults.push(r);
    console.log(`Reqs: ${r.total_requests}, RPS: ${r.throughput_rps}, Avg: ${r.avg_ms}ms, P95: ${r.p95_ms}ms, P99: ${r.p99_ms}ms, Errors: ${r.errors}`);
  }

  const file = `load_test_${label}.json`;
  fs.writeFileSync(file, JSON.stringify(allResults, null, 2));
  console.log(`\nResults saved to ${file}`);
}

main().catch(console.error);
